Add numpy-style dtypes to Variable factories.

gchanan opened the pull request:
1) Add numpy-style dtypes corresponding to torch tensor types.  These are:
torch.float16, torch.float32, torch.float64, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64
as well as torch.cuda, torch.sparse, and torch.cuda.sparse equivalents.

2) Adds "legacy" names for the above dtypes that correspond more closely to existing tensor names.  These are:
torch.half, torch.float, torch.double, torch.short, torch.int, torch.long.
torch.byte and torch.char don't exist because they either don't match numpy semantics or differ on different architectures.

3) Adds a "dtype" parameter to Variable factories (e.g. zeros, ones) that allows the user to specify the type without changing the default tensor type.

4) Adds a "dtype" getter to Variables that return the canonical dtype from 1)

This PR is missing the following useful features that should be added in the future:
A) We only add the "dtype" parameter to auto-generated factories; hand-written factories like in tensor_new.cpp don't support this yet.

B) We don't allow type conversions to use dtypes; that should be added to type(param) or a new function.

C) We don't yet have a "device" parameter for these factories; right now, they will only create Variables on the default device.

gchanan committed:
Add numpy-style dtypes to Variable factories.

1) Add numpy-style dtypes corresponding to torch tensor types.  These are:
torch.float16, torch.float32, torch.float64, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64
as well as torch.cuda, torch.sparse, and torch.cuda.sparse equivalents.

2) Adds "legacy" names for the above dtypes that correspond more closely to existing tensor names.  These are:
torch.half, torch.float, torch.double, torch.short, torch.int, torch.long.
torch.byte and torch.char don't exist because they either don't match numpy semantics or differ on different architectures.

3) Adds a "dtype" parameter to Variable factories (e.g. zeros, ones) that allows the user to specify the type without changing the default tensor type.

4) Adds a "dtype" getter to Variables that return the canonical dtype from 1)

This PR is missing the following useful features that should be added in the future:
A) We only add the "dtype" parameter to auto-generated factories; hand-written factories like in tensor_new.cpp don't support this yet.

B) We don't allow type conversions to use dtypes; that should be added to type(param) or a new function.

C) We don't yet have a "device" parameter for these factories; right now, they will only create Variables on the default device.

gchanan committed:
backend_to_string can be private.

gchanan committed:
Define python binding argument indexes in a more simple way.

gchanan commented:
The issue here appears to be if you compile with NO_CUDA=1, you don't define the cuda dtypes, even though as of now you get the torch.cuda.XTensor types.  I believe https://github.com/pytorch/pytorch/pull/5225 has the same issue, although maybe it isn't exposed by the existing tests (we should test it, though).

CC @colesbury.

gchanan committed:
add all_declared_types, still need to hook it up to THPDType.

gchanan committed:
Fix all_declared_types for missing types (it's Sparse + Half).

gchanan committed:
Ensure cuda dtypes are created even if compiled with NO_CUDA=1.

ezyang commented:
Looks nice. A few questions:

1. The intent is that once Variable-Tensor merge, then the Variable factories will replace `torch.zeros`, is that right?
2. Suppose I create a Variable using `_VariableFunctions.zeros`. What is the idiomatic way of creating a tensor that also requires gradient? Will we add a `requires_grad` flag to these functions? It kind of looks like `gen_python_functions.py` is handling this case but I don't see tests exercising it.

gchanan commented:
1. Yes.
2. That already exists.  You are correct that there aren't tests that exercise both simultaneously though.

gchanan committed:
Fix case where dtype is provided but dispatch is via namespace.

This happens in ones_like, empty_like, randn_like.

There is some question if we should do:
1) at::ones_like(tensor).toType(dtype)
2) at::ones_like(tensor.toType(dtype))

I did the former because this matches with the numpy documentation, i.e.:
"Overrides the data type of the result." and it's easier to implement.

Note that the above causes an extra copy, either of the input or output.
Here's a better implementation:
1) Make zeros_like, ones_like native functions that take an optional type (named dtype?).
2) Match the type argument with the dtype, so we don't have two different parameters.
3) Call at::zeros_like(input, type) -> at::native::zeros_like(input, type) -> type.zeros(input.sizes())

colesbury commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -431,6 +431,14 @@ PyObject *THPVariable_is_sparse(THPVariable *self)
+{
+  HANDLE_TH_ERRORS
+  auto& self_ = self->cdata;
+  return torch::autograd::utils::wrap((THPDtype*)torch::getDtype(self_.type()));
```
The `THPDtype*` cast looks redundant.

colesbury commented on torch/csrc/DynamicTypes.cpp:
```diff
@@ -100,6 +105,14 @@ at::Type& getATenType(PyTypeObject* type)
+  if (it != attype_to_dtype.end()) {
+    return it->second;
+  }
+  throw std::invalid_argument("unsupported at::Type");
```
Including the name of the function will help debugging in case this is actually hit. e.g.:

```
getDtype(): unsupported at::Type
```

colesbury commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -218,6 +223,22 @@ inline std::vector<int64_t> PythonArgs::intlistWithDefault(int i, std::vector<in
+inline const at::Type& PythonArgs::dtype(int i) {
+  if (!args[i]) return default_type();
+  PyObject *arg = args[i];
+  if (!THPDtype_Check(arg)) {
```
Don't double-check the argument type. It's already checked by the parser.

gchanan replied:
sure; fwiw this seems to be done often in these functions (see e.g. Generator).

colesbury commented on torch/csrc/Dtype.h:
```diff
@@ -0,0 +1,25 @@
+
+extern PyObject *THPDtypeClass;
+
+#define THPDtype_Check(obj) ((PyObject*)Py_TYPE(obj) == THPDtypeClass)
```
Prefer inline functions to macros.

apaszke commented:
NB: I don't think that creating a tensor using one of those factories that also requires gradient is an often use case

apaszke commented on test/test_torch.py:
```diff
@@ -1079,6 +1079,25 @@ def test_ones_like_multiple_device(self):
+                out = torch._C._VariableFunctions.eye(3, 3, dtype=dtype)
+                self.assertEqual(out.dtype, dtype)
+            self.assertEqual(dtype in cuda_dtypes, dtype.is_cuda)
+            self.assertFalse(dtype.is_sparse)
```
Can we deduplicate this test into a generic thing, and call that from `test_sparse`?

gchanan replied:
I think so?  I'm not that familiar with test_sparse, but it seems possible.

colesbury commented on torch/csrc/Dtype.h:
```diff
@@ -0,0 +1,25 @@
+
+PyObject * THPDtype_New(at::Type* cdata, const std::string& name, bool is_cuda, bool is_sparse);
+
+#ifdef _THP_CORE
```
Don't do this stuff. The THP.h/_THP_CORE stuff is a mess. Peter's torch.h should be the preferred C++ interface for PYTorhc.

apaszke commented on torch/csrc/utils/tensor_dtypes.cpp:
```diff
@@ -0,0 +1,93 @@
+
+namespace torch { namespace utils {
+
+static std::pair<std::string, std::string> getDtypeNames(at::ScalarType scalarType) {
```
Is there a reason why we don't want to make those `const char*`? Remember that this will create copies of both strings, irrespectively of what you'll be doing with them (not a big deal if this is only used in initialization though).

gchanan replied:
I just did it that way because it's only in initialization and std::strings are nicer to use.  I can change it if you want.

colesbury commented on torch/csrc/utils/tensor_dtypes.cpp:
```diff
@@ -0,0 +1,93 @@
+  auto torch_module = THPObjectPtr(PyImport_ImportModule("torch"));
+  auto cuda_module = THPObjectPtr(PyImport_ImportModule("torch.cuda"));
+  auto sparse_module = THPObjectPtr(PyImport_ImportModule("torch.sparse"));
+  auto cuda_sparse_module = THPObjectPtr(PyImport_ImportModule("torch.cuda.sparse"));
```
You should check for null (and throw `python_error()`) after these calls.

colesbury commented on torch/csrc/utils/tensor_dtypes.cpp:
```diff
@@ -0,0 +1,93 @@
+    if (type) {
+      torch::registerDtypeObject(dtype, *type);
+    }
+    PyModule_AddObject(module, primary_name.c_str(), (PyObject*)dtype);
```
Check the return value (`int`) for errors

colesbury replied:
Also, I think you need to `Py_INCREF` dtype here.

colesbury commented on torch/csrc/DynamicTypes.h:
```diff
@@ -19,10 +20,13 @@ void registerStoragePyTypeObject(
 PyObject* createPyObject(const at::Storage& storage);
 PyTypeObject* getPyTypeObject(const at::Tensor& tensor);
 at::Type& getATenType(PyTypeObject* type);
+THPDtype* getDtype(at::Type& type);
```
`const at::Type&`?

colesbury commented on torch/csrc/utils/tensor_dtypes.cpp:
```diff
@@ -0,0 +1,93 @@
+    }
+    PyModule_AddObject(module, primary_name.c_str(), (PyObject*)dtype);
+    if (legacy_name != "") {
+      PyModule_AddObject(module, legacy_name.c_str(), (PyObject*)dtype);
```
and `Py_INCREF` here

colesbury commented on torch/csrc/utils/tensor_dtypes.cpp:
```diff
@@ -0,0 +1,93 @@
+      }
+      default: throw std::runtime_error("Unimplemented backend");
+    }
+    auto baseType = context.type_registry[static_cast<int>(backend)][static_cast<int>(scalarType)].get();
```
Might be worth a comment as to why you're not using `Context::getType(...)` here.

colesbury commented on tools/autograd/templates/python_torch_functions_dispatch.h:
```diff
@@ -29,6 +29,23 @@ static at::Type& default_type() {
+}
+
+static const at::Type& maybe_initialize_cuda(const at::Type &type) {
+  if(type.is_cuda()) {
```
`if (`

gchanan commented on tools/autograd/templates/python_torch_functions_dispatch.h:
```diff
@@ -29,6 +29,23 @@ static at::Type& default_type() {
+  });
+}
+
+static const at::Type& maybe_initialize_cuda(const at::Type &type) {
```
because at one point, I was doing the initialize + return inline, i.e.
`
maybe_initialize_cuda(type).foobar(...)
`

but this didn't jive well with how the codegen was written, so I removed it.  I'll remove the return type.

gchanan committed:
Don't return from maybe_initialize_cuda.

gchanan committed:
Don't leak DType name.

gchanan committed:
Address cpp review comments.

gchanan committed:
Share code between sparse and non-sparse test_dtypes.

gchanan committed:
Rewrite _like functions as native function with explicit type parameter.

colesbury commented on torch/csrc/Dtype.h:
```diff
@@ -0,0 +1,24 @@
+  bool is_sparse;
+};
+
+extern PyObject *THPDtypeClass;
```
This seems unnecessary to me. Use `extern PyTypeObject* THPDtypeType` instead. Then you don't have to worry about initializing it.

The reason we use the separate class/Type pattern in other places is because we subclass `_VariableBase` with `Variable` in Python, but want to construct the subclass from C++. Since there's no subclass of THPDtypeType here it doesn't matter.

gchanan committed:
Use type 'Type' instead of 'dtype' for consistency.

colesbury commented on torch/csrc/utils/tensor_dtypes.cpp:
```diff
@@ -0,0 +1,104 @@
+      torch::registerDtypeObject((THPDtype*)dtype, *type);
+    }
+    Py_INCREF(dtype);
+    if (PyModule_AddObject(module, primary_name.c_str(), (PyObject*)dtype) != 0) {
```
These `(PyObject*)` casts look unnecessary now

colesbury reviewed (approved):
See comments inline. Otherwise looks good to me

gchanan commented:
CC @ezyang: this now supports Type as parameters to native functions, which will get turned into dtypes at the python level (assuming the function is bound).

gchanan committed:
Address review comments.

gchanan committed:
Handle arg_idx when there is requires_grad but no dtype in python_binding_arguments.

yf225 commented:
@pytorchbot retest this please
